laurent22/joplin
# Chore: Mobile: Migrate `NoteItem` and `Checkbox` to TypeScript

# Summary

This pull request migrates `note-item.js` and `checkbox.js` to TypeScript.

**Additional changes**:
- `NoteItem` was made to extend [PureComponent](https://react.dev/reference/react/PureComponent) because its rendered output is completely determined by its `state` and `props`.
- `Checkbox` was converted to a function component.

## Style comparison

**Note**: Web screenshots are from a branch with style fixes from https://github.com/laurent22/joplin/pull/11071 applied (fixes the scrollbar).

**Web/light mode**:

![two screenshots: before and after: Appearance of the note list seems the same](https://github.com/user-attachments/assets/1e48ff93-1613-444d-9ced-0c28748a7f15)

**Web/dark mode**

![before/after comparison screenshots: Like light mode screenshot, but dark mode](https://github.com/user-attachments/assets/cf4e791a-c956-426f-adad-13e03d8f34ee)



# Testing plan

1. Open a folder with many notes, at least 3 of them to-dos.
2. Mark one of the to-dos as completed.
3. Switch to a different folder.
4. Switch back to the folder from step 1.
5. Verify that the to-do is still marked as completed.
6. Open the to-do in the note editor.
7. Mark the to-do as incomplete.
8. Go back to the note list.
9. Verify that the to-do is marked as incomplete.

This has been tested successfully on:
- [x] Web+Chromium
- [x] Android 13 (in progress)


Additional Android testing:
1. Enable TalkBack
2. Move TalkBack focus to the first to-do. Verify that TalkBack reads the note title.
3. Move to the next item.
4. Verify that the note's checkbox is focused. Verify that TalkBack reads the checkbox's state (checked or not checked) and the note title.
5. Double-tap.
6. Verify that TalkBack reads the checkbox's new state.
7. Verify that the to-do is has been toggled (for me, the item is marked as complete).
